File: ios/LodoCore/Sources/LodoCore/ContactGraphLayout.swift
import Foundation

/// 关系图谱(ContactGraphView)的纯布局数学:不依赖 SwiftUI/Canvas,放 LodoCore
/// 才能脱离模拟器单测。用元组而不是 CGPoint,保持这个包"无 UI 依赖"的约束——
/// 调用处(app 层)自己转成 CGPoint。
public enum ContactGraphLayout {
    /// 把 count 个节点均匀摆在以 center 为圆心、radius 为半径的圆周上,从
    /// 12 点钟方向顺时针分布。count <= 1 时把唯一的节点(如果有)放在圆心,
    /// 不需要摆出一个圆——一个点谈不上"分布"。
    public static func circlePositions(
        count: Int, radius: Double, center: (x: Double, y: Double)
    ) -> [(x: Double, y: Double)] {
        guard count > 0 else { return [] }
        guard count > 1 else { return [center] }
        return (0..<count).map { index in
            let angle = 2 * Double.pi * Double(index) / Double(count) - .pi / 2
            return (
                x: center.x + radius * cos(angle),
                y: center.y + radius * sin(angle)
            )
        }
    }
}

File: android/app/src/main/java/com/lodo/app/core/ContactGraphLayout.kt
package com.lodo.app.core

import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.sin

/**
 * 人脉关系图谱的确定性圆形布局,与 iOS ContactGraphLayout.circlePositions 对齐:
 * 不做力导向仿真,N 个节点均匀布在一个圆上,结果稳定可预测、可单测。纯
 * Kotlin/JVM,不依赖 Compose 的 Offset 类型,UI 层转换成自己的坐标类型。
 */
object ContactGraphLayout {
    /** 返回每个节点相对圆心 (0,0) 的 (x, y) 坐标,半径为 radius;从正上方开始
     * 顺时针均匀分布。count <= 0 返回空;count == 1 时该节点在圆心。 */
    fun circlePositions(count: Int, radius: Double): List<Pair<Double, Double>> {
        if (count <= 0) return emptyList()
        if (count == 1) return listOf(0.0 to 0.0)
        val step = 2 * PI / count
        return (0 until count).map { i ->
            val angle = -PI / 2 + i * step
            radius * cos(angle) to radius * sin(angle)
        }
    }
}
